### PR TITLE
Pubsub + DNS Facts

### PR DESCRIPTION
**Pull request by rambleraptor**
This is Pubsub Facts + some additional facts functionality to help build more of these

* Pubsub Facts
* Define which key is used to read the returned list (`list_key`)
* Turn off filters for list APIs that do not support filtering
* Plumb facts values back into the original object (integration tests are done as part of the original module creation, but require information about the facts modules)
* Custom tests (instead of just checking that a single thing exists)
* DNS facts

TLDR; The DNS facts should be the messiest + require the most custom stuff. I'm feeling pretty confident that most of the facts modules should be able to use the same custom flags that Pubsub + DNS use.

-----------------------------------------------------------------
# [all]

### CODE REVIEW

File: templates/ansible/facts.erb [hunk 1]
```diff
@@ -103,16 +118,21 @@ def main():
     module.exit_json(**return_value)
 
 
+<% if object.collection -%>
+<%= lines(emit_link('collection', build_url(object.collection), object)) -%>
```
Comment (danawillow) [inline templates/ansible/facts.erb]: Is there a way that this could have been some sort of override instead of a separate field?

File: templates/ansible/verifiers/facts.yaml.erb [hunk 2]
```diff
@@ -5,10 +5,19 @@
 -%>
 - name: verify that <%= object.name.underscore -%> was <%= lines(verbs[_state.to_sym]) -%>
   <%= module_name -%>:
+<% if object.facts.has_filters && object.facts.filter.name == 'filters' -%>
```
Comment (danawillow) [inline templates/ansible/verifiers/facts.yaml.erb]: what if it has filters but it isn't called 'filters'? is that ignored on purpose?

File: templates/ansible/verifiers/facts.yaml.erb [hunk 3]
```diff
@@ -5,10 +5,19 @@
...
       filters:
          - name = <%= name_parameter %>
+<% end #if object.facts.has_filters -%>
 <% @parameters.each do |k, v| -%>
-      <%= k -%>: <%= lines(v) -%>
+<%
+  value = if @parameters.keys.first == k && v.include?('{{')
```
Comment (danawillow) [inline templates/ansible/verifiers/facts.yaml.erb]: what's the significance of the first key?

File: provider/ansible.rb [hunk 4]
```diff
@@ -291,6 +298,14 @@ def compile_datasource(data)
         )
       end
 
+      def datasource_info(data)
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/ansible.rb]: Yeah good call. Didn't do a great job commenting this + explaining. Adding a bunch of new comments and changing old ones!

File: templates/ansible/facts.erb (hunk 1 above)
Comment (rambleraptor) [reply templates/ansible/facts.erb]: I definitely think it should be! Currently, Puppet, Chef, Ansible all handle collection URLs this way. I'd definitely like to make that change, but maybe wait for a different PR?

File: templates/ansible/verifiers/facts.yaml.erb (hunk 2 above)
Comment (rambleraptor) [reply templates/ansible/verifiers/facts.yaml.erb]: There's two different cases:

1) GCE (the 'filters' notation)
2) Everything else (that I've seen so far)

I added some comments and a .is_gce? function to make this more clear.

File: provider/ansible/facts_override.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  end
+end
+
+# This is a property exclusive to Ansible filters.
+# This is the default property used for filter information on Ansible.
+# By using Api::Types, we get more flexibility and a lot for free.
+class FilterProp < Api::Type::Array
+  def validate
+    @item_type ||= 'Api::Type::String'
+    @name ||= 'filters'
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/ansible/facts_override.rb]: Definitely!

File: provider/ansible.rb [hunk 6]
```diff
@@ -240,6 +242,11 @@ def get_example(cfg_file)
       end
 
       def generate_resource(data)
+        # Add facts from datasources to non-datasource if they exist.
+        # Some values from datasources.facts may be necessary for building integration
+        # tests, which is done before datasource building.
+        # We don't want to duplicate those .facts values though.
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/ansible.rb]: Yeah, this wasn't well written. Fixed!

File: provider/ansible/resource_override.rb [hunk 7]
```diff
@@ -33,12 +34,15 @@ module OverrideProperties
...
       attr_reader :version_added
+
+      attr_reader :facts
     end
 
     # Product specific overriden properties for Ansible
     class ResourceOverride < Provider::ResourceOverride
       include OverrideProperties
       # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics/MethodLength
```
[the remark this answers is not in the document]
Comment (rambleraptor) [reply provider/ansible/resource_override.rb]: agreed!

File: templates/ansible/facts.erb (hunk 1 above)
Comment (rambleraptor) [reply templates/ansible/facts.erb]: Filing an issue for this: https://github.com/GoogleCloudPlatform/magic-modules/issues/453

**Comment by rambleraptor** [comment]
@danawillow ptal!

File: provider/ansible/example.rb [hunk 8]
```diff
@@ -206,15 +233,14 @@ def build_parameters(object)
...
-              .map(&:name)
-              .select { |para| url_parts.include? para }
-              .map { |para| { para => sample_code[para] } }
-              .reduce({}, :merge)
+        parameters = handwritten_vals_for_properties(object,
+                                                     uri_properties(object, ignored_props))
+
+        # Grab values for filters.
+        underscore_name = object.facts.filter.name.underscore
+        parameters[underscore_name] = sample_code[underscore_name] if sample_code[underscore_name]
```
Comment (danawillow) [inline provider/ansible/example.rb]: do you actually need this 'if' given that you have the compact after it?
